### PR TITLE
Fix crash when editableText is not mounted

### PR DESCRIPTION
**Pull request by matasb-google**
Fix issue when text_selection crashes due to editableText not being mounted.

No issue filled (internal b/425840577)

Likely introduced by https://github.com/flutter/flutter/pull/167881

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I signed the [CLA].
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I added new tests to check the change I am making, or this PR is [test-exempt].
- [X] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [X] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

**Review by matasb-google — COMMENTED** [review]
<removed>

**Review by matasb-google — COMMENTED** [review]
<removed>

**Comment by matasb-google** [comment]
@Renzo-Olivares  - could you take a look?

File: packages/flutter/lib/src/widgets/text_selection.dart [hunk 1]
```diff
@@ -2929,6 +2929,9 @@ class TextSelectionGestureDetectorBuilder {
   }
 
   void _onSingleLongTapEndOrCancel() {
+    if (!_isEditableTextMounted) {
```
Comment (Renzo-Olivares) [inline packages/flutter/lib/src/widgets/text_selection.dart]: I think we can make `_isEditableTextMounted` the first condition in the "Update the floating cursor" logic below, this way we still reset some internal state when the long tap is cancelled.

**Comment by Renzo-Olivares** [comment]
This will need a test before it can be merged.

File: packages/flutter/lib/src/widgets/text_selection.dart [hunk 2]
```diff
@@ -2934,6 +2934,7 @@ class TextSelectionGestureDetectorBuilder {
     _dragStartViewportOffset = 0.0;
     _dragStartScrollOffset = 0.0;
     if (defaultTargetPlatform == TargetPlatform.iOS &&
+        _isEditableTextMounted &&
```
Comment (Renzo-Olivares) [inline packages/flutter/lib/src/widgets/text_selection.dart]: super nit: let `_isEditableTextMounted` be at the front of the logic in this if branch.

**Comment by matasb-google** [comment]
> This will need a test before it can be merged.

I do not see good way to test this. This is edge case, with very specific edge case. 
As it is bug fix for crash, I would expect this to be acceptable without specific test?

**Comment by matasb-google** [comment]
Friendly request to review and approve. Tests are not feasible, did not find way to reproduce using tests. This is crash at Google/money client, and we would like to fix in github, instead of workarounds internally.

**Review by Renzo-Olivares — APPROVED** [review]
LGTM, sorry for the delay. I also tried writing a test but could not reproduce the bug (`editableText` was always mounted for me).

**Comment by Renzo-Olivares** [comment]
I'm going to request a test exemption on discord and then we can land this.

**Comment by Renzo-Olivares** [comment]
Hi @matasb-google, this PR doesn't qualify for a test exemption because it does not fall under the criteria noted in https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests. In most cases when fixing a regression like this one we want to add a test to make sure we don't break the fix in the future. This particular fix was rather tricky to write a test for, but I gave it another shot and was able to trigger the crash. If you add this test to the PR then we can move forward and land it.

```dart
testWidgets(
  'does not crash when long press is cancelled after unmounting',
  (WidgetTester tester) async {
    // Regression test for b/425840577.
    final ScrollController scrollController = ScrollController();
    addTearDown(scrollController.dispose);

    await tester.pumpWidget(
      MaterialApp(
        home: Material(
          child: CustomScrollView(
            controller: scrollController,
            slivers: <Widget>[
              SliverList(
                delegate: SliverChildBuilderDelegate(
                  (BuildContext context, int index) {
                    if (index == 0) {
                      return const TextField();
                    }
                    return const SizedBox(height: 50);
                  },
                  childCount: 200,
                  addAutomaticKeepAlives: false,
                ),
              ),
            ],
          ),
        ),
      ),
    );

    final EditableTextState state = tester.state<EditableTextState>(find.byType(EditableText));
    // Start a long press, don't release it, and don't completely reach kLongPressTimeout so the
    // gesture is not accepted and is cancelled when the recognizer is disposed.
    await tester.startGesture(tester.getCenter(find.byType(TextField)));
    await tester.pump(const Duration(milliseconds: 200));
    await tester.pumpAndSettle();

    // While attempting to long press, scroll the TextField out of view
    // to dispose of it and its gesture recognizers.
    scrollController.jumpTo(8000.0);
    await tester.pump();
    expect(state.mounted, isFalse);
    // Should reach the end of the test without any failures.
  },
  variant: TargetPlatformVariant.only(TargetPlatform.iOS),
);
```

**Comment by matasb-google** [comment]
Thanks, added test.

**Review by loic-sharma — APPROVED** [review]
Looks good to me, thanks for taking time to create a test for this!
